#!/bin/sh
set -eu

if ! command -V patch >/dev/null 2>&1; then
  extra_root="$(readlink -f "$(dirname -- "$0")/..")"
  "${extra_root}/bin/pkg" add patch
fi

generator="$(find /usr/lib/python*.*/site-packages/batocera_launch/emulators -name cemu.py)"
patch -b -z .backup-cemuhook "${generator}" << 'EOF'
--- "${generator}"
+++ cemu.py
@@ -27,6 +27,7 @@
 _WIIMOTE: Final = 'Wiimote'
 
 _API_SDL: Final = 'SDLController'
+_API_DSU: Final = 'DSUController'
 _API_WIIMOTE: Final = 'Wiimote'
 
 # from https://github.com/cemu-project/Cemu/blob/main/src/input/emulated/WPADController.h
@@ -43,6 +44,8 @@
 _WIIMOTE_NAME_NUNCHUK: Final = f'{_WIIMOTE_NAME} Nunchuk'
 _WIIMOTE_NAME_CLASSIC: Final = f'{_WIIMOTE_NAME} Classic Controller'
 
+_DEFAULT_IP: Final = 'localhost'
+_DEFAULT_PORT: Final = '26760'
 _DEFAULT_DEADZONE: Final = '0.25'
 _DEFAULT_RANGE: Final = '1'
 
@@ -531,6 +534,15 @@
                 ET.SubElement(entry_node, 'mapping').text = key
                 ET.SubElement(entry_node, 'button').text = value
 
+            # Add motion control
+            if 'cemuhook_server_port' in self.config:
+                motion_controller_node = ET.SubElement(root, 'controller')
+                ET.SubElement(motion_controller_node, 'api').text = _API_DSU
+                ET.SubElement(motion_controller_node, 'uuid').text = '0'
+                ET.SubElement(motion_controller_node, 'motion').text = 'true'
+                ET.SubElement(motion_controller_node, 'ip').text = self.config.get_str('cemuhook_server_ip', _DEFAULT_IP)
+                ET.SubElement(motion_controller_node, 'port').text = self.config.get_str('cemuhook_server_port', _DEFAULT_PORT)
+
             # Save to file
             with (self.controller_profiles_dir / f'controller{nplayer}.xml').open('wb') as handle:
                 tree = ET.ElementTree(root)
EOF

echo >&2 "Cemu generator was patched for cemuhook."
